### PR TITLE
chore: refactor API tokens

### PR DESCRIPTION
**Pull request by netroms**

### CODE REVIEW

File: dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/security/apikey/ApiTokenServiceImplTest.java [hunk 1]
```diff
@@ -215,4 +219,33 @@ private void switchToOtherUser()
...
+
+    @Test
+    void testHashingToken()
+    {
+        char[] token = ApiTokenServiceImpl.generatePlainTextToken( ApiTokenType.PERSONAL_ACCESS_TOKEN_V1 );
+        String hashedToken = hashToken( token );
+        assertTrue( isValidSHA256( hashedToken ) );
+    }
+
+    private static boolean isValidSHA256( String s )
```
Comment (david-mackessy) [inline dhis-2/dhis-test-integration/src/test/java/org/hisp/dhis/security/apikey/ApiTokenServiceImplTest.java:238]: if this logic is used to assert in a test, should it actually be in the src code itself (not in a test) with it's own test?
for instance, this is just validating test instances of a valid hash. Should this also be done in the code?

**Review by david-mackessy — APPROVED** [review]

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/security/apikey/ApiToken.java [hunk 2]
```diff
@@ -44,86 +48,46 @@
 /**
  * @author Morten Svanæs <msvanaes@dhis2.org>
  */
+@Getter
+@Setter
+@Builder( toBuilder = true )
 @JacksonXmlRootElement( localName = "apiToken", namespace = DxfNamespaces.DXF_2_0 )
-public class ApiToken extends BaseIdentifiableObject implements MetadataObject
+public final class ApiToken extends BaseIdentifiableObject implements MetadataObject
```
Comment (jbee) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/security/apikey/ApiToken.java:55]: I am wondering if `final` on a hibernate mapped class could pose an issue for the hibernate proxy.

File: dhis-2/dhis-api/src/main/java/org/hisp/dhis/security/apikey/ApiTokenService.java [hunk 3]
```diff
@@ -29,32 +29,38 @@
...
 
-    ApiToken getWithUid( String uid );
+    @Nonnull
+    List<ApiToken> getAll();
+
+    @Nonnull
+    List<ApiToken> getAllOwning( @Nonnull User user );
+
+    @CheckForNull
+    ApiToken getWithUid( @Nonnull String uid );
```
Comment (jbee) [inline dhis-2/dhis-api/src/main/java/org/hisp/dhis/security/apikey/ApiTokenService.java:56]: I think we use `by` as in `getByUid` name convention elsewhere.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/security/apikey/ApiTokenResolver.java [hunk 4]
```diff
@@ -27,108 +27,156 @@
...
+        return false;
+    }
+
+    private static void checksum( char[] token )
+    {
+        try
+        {
+            if ( !validateChecksum( token ) )
+            {
+                throw new ApiTokenAuthenticationException( ApiTokenErrors
```
Comment (jbee) [inline dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/security/apikey/ApiTokenResolver.java:155]: Always hard to say if exceptions are only used within the service layer but never meant to leave it. Just want to emphasis that if this is not caught and rethrown as one of our feedback exceptions we should use those here instead.

**Review by jbee — COMMENTED** [review]
Couple of comments and questions.

File: dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/security/ApiTokenController.java [hunk 5]
```diff
@@ -159,26 +136,20 @@ public WebMessage postJsonObject( HttpServletRequest request )
...
-    {
-        newToken.setKey( oldToken.getKey() );
-        validateApiKeyAttributes( newToken );
-    }
-
-    protected void validateBeforeCreate( ApiToken token )
+    @PostMapping( consumes = { "application/xml", "text/xml" } )
+    @ResponseBody
+    public WebMessage postXmlObject( HttpServletRequest request )
+        throws HttpRequestMethodNotSupportedException
```
[the remark this answers is not in the document]
Comment (netroms) [reply dhis-2/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/security/ApiTokenController.java:142]: Yes, that was my intention.

**Review by david-mackessy — APPROVED** [review]

**Review by jbee — APPROVED** [review]

**Pull request merged**